tests/ui/specialization/min_specialization/bad-const-wf-doesnt-specialize.rs

```diff
@@ -0,0 +1,38 @@
+#![feature(min_specialization)]
+
+// An impl that has an erroneous const substitution should not specialize one
+// that is well-formed.
+//
+// This is what used to happen:
+//
+// ## IMPL (1.) -- `impl<const N: i32> Copy for S<N> {}`
+//
+// When checking if this impl specializes the next one, we take the trait ref
+// and unify it against impl #2's trait ref with fresh vars. The trait ref of
+// impl #1 is `S<Const(Param(N/#0): usize)>` because of astconv using the type
+// from the `L` substitution of struct `S`. This unifies successfully with impl #2,
```

Comment:
> The trait ref of impl #1 is `S<Const(Param(N/#0): usize)>` because of astconv using the type from the `L` substitution of struct `S`.

I guess I could be a bit more explicit here -- I'm talking about this code:

https://github.com/rust-lang/rust/blob/39f2657d1101b50f9b71ae460b762d330cc8426b/compiler/rustc_hir_analysis/src/astconv/mod.rs#L449-L458

So when instantiating the substs for `S<N>`, we use the const param did from `L`, and end up with `S<Const(Param(N/#0): usize)>`, even though `N` has the type `i32` at its declaration. This is surprising, I guess, but also makes sense from the implementation of the astconv code.